### PR TITLE
v0.2.0: Field/FieldCt lift, CT siblings, basic/constrained/strict mod…

### PR DESCRIPTION
**Pull request by kaidokert**
…ule reorg

Major changes:

- Adds modmath::Field<T> and modmath::FieldCt<T> with lifetime-branded Residue<'f, T> / ResidueCt<'f, T> as the canonical Montgomery surface.
- Adds CT-finalize siblings throughout the Montgomery stack: wide_redc_ct, cios_montgomery_mul_ct, wide_montgomery_mul_ct, basic_montgomery_mod_exp_ct, plus _pr variants where pre-reduced inputs are assumed.
- Adds CiosMontMulCt convenience trait paralleling CiosMontMul.
- Adds FieldCt::exp_public_exp for the CT-base + NCT-exp shape (RSA encrypt with public e, Curve25519 Fermat inverse with public p-2).
- Reorganizes schoolbook surface: modmath::basic, modmath::constrained, modmath::strict modules replace the basic_*/constrained_*/strict_* function prefixes; backward-compat re-exports preserved.
- Adds _pr (pre-reduced) sibling functions to schoolbook surface so CT-aware code paths can drop the Rem bound.
- wide-mul is now a default feature.
- Subtle dep added (gated by wide-mul which pulls fixed-bigint/ct).

## Summary by Sourcery

Introduce a branded Montgomery field abstraction with CT/NCT personalities, reorganize the public modular arithmetic surface into basic/constrained/strict modules, and extend Montgomery and schoolbook operations with pre-reduced and constant-time variants backed by wide-REDC/CIOS.

New Features:
- Add Field<T, P> / FieldCt / FieldNct Montgomery context types with lifetime-branded Residue values and personality parameter for CT vs non-CT arithmetic.
- Expose basic/constrained/strict module hubs that group the schoolbook add/sub/mul/exp/inv and montgomery APIs, including wide-REDC helpers and CT submodules.
- Provide constant-time wide-REDC primitives (wide_redc_ct, wide_montgomery_mul_ct) and CIOS-based Montgomery multiplication (cios_montgomery_mul_ct) with a CiosMontMulCt trait.
- Add constant-time Montgomery exponentiation helpers in the basic flavor, including pre-reduced and CT-over-exponent variants and FieldCt::exp_public_exp for public exponents.
- Introduce pre-reduced (_pr) variants of add/sub/mul/exp and Montgomery conversion/multiply/exp functions across basic, constrained, and strict surfaces that operate without Rem bounds.

Enhancements:
- Clarify and document NPrimeMethod behavior, R>N vs R=2^W semantics, and wide-REDC usage throughout the Montgomery stack.
- Refactor montgomery module layout to keep flavor-neutral items at modmath::montgomery while routing R>N paths through basic/constrained/strict::montgomery, and remove wide_from_montgomery in favor of wide_redc.
- Tighten and document interaction with fixed-bigint personality typestate so CT paths require Ct-typed integers and Rem/Div are only available on Nct personalities.
- Improve schoolbook modular add/sub/mul/exp implementations by factoring shared logic into pre-reduced helpers, handling modulus==1 in exponentiation, and reducing trait bounds for non-Rem callers.
- Add reference-based wide-REDC helpers (strict_wide_redc, strict_wide_montgomery_mul and their CT siblings) for non-Copy bigint backends and relax Copy constraints where unnecessary.
- Expand and rewrite crate-level and module-level documentation to describe the reorganized API surface, personalities, and CT vs non-CT guarantees.

Build:
- Bump crate version from 0.1.5 to 0.2.0 and update fixed-bigint to 0.3.
- Add subtle as a dependency for constant-time conditional selection and equality, and make wide-mul a default feature with explicit docs.rs feature configuration.

Documentation:
- Update top-level crate docs and Montgomery module docs to describe the new basic/constrained/strict structure, Field/Residue types, wide-REDC/CIOS usage, and CT vs non-CT variants.
- Add extensive inline documentation and examples around fixed-bigint personalities, pre-reduced APIs, and constant-time behavior assumptions.

Tests:
- Add comprehensive tests for pre-reduced add/sub/mul/exp operations over fixed-bigint types.
- Introduce equivalence tests ensuring CT wide-REDC and CIOS variants match their variable-time counterparts across small integers and FixedUInt backends.
- Add tests for Field/Residue behavior, including round-trips, CT vs NCT cross-checks, cswap, exp_public_exp equivalence, branding limitations, and from_precomputed construction.
- Extend existing Montgomery and exponentiation test suites to cover fixed-bigint personalities, pre-reduced paths, and modulus==1 edge cases.

Chores:
- Add or update repository ignore rules and adjust internal comments to reflect the removal of legacy terminology and TODOs around the R>N path.

## Summary by CodeRabbit

* **New Features**
  * Public Montgomery Field/Residue API, new flavor hubs (basic, constrained, strict), pre-reduced entry points, wide/strict Montgomery facades, and constant-time variants.

* **Changes**
  * Crate reorganized and documented, public surface streamlined, default features adjusted (wide-mul enabled), version bumped to 0.2.0.

* **Tests**
  * Expanded coverage for pre-reduced paths, Montgomery/wide-REDC, constant-time equivalence, and bigint/personality cases.

* **Chores**
  * Ignore macOS .DS_Store files.

### CODE REVIEW

File: modmath/src/constrained/montgomery.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#[doc(inline)]
+pub use crate::montgomery::constrained_mont::constrained_montgomery_mul as mul;
+#[doc(inline)]
+pub use crate::montgomery::constrained_mont::constrained_montgomery_mod_mul as mod_mul;
+#[doc(inline)]
+pub use crate::montgomery::constrained_mont::constrained_montgomery_mod_mul_with_method as mod_mul_with_method;
+#[doc(inline)]
+pub use crate::montgomery::constrained_mont::constrained_montgomery_mod_exp as mod_exp;
+#[doc(inline)]
+pub use crate::montgomery::constrained_mont::constrained_montgomery_mod_exp_with_method as mod_exp_with_method;
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/constrained/montgomery.rs]: done

File: modmath/src/field.rs [hunk 2]
```diff
@@ -0,0 +1,1018 @@
...
+    ///
+    /// The squarings and multiplications themselves go through CT primitives
+    /// ([`cios_montgomery_mul_ct`](crate::montgomery::cios::cios_montgomery_mul_ct)),
+    /// so the
+    /// base and intermediate Montgomery values do not leak through timing.
+    /// What DOES leak is the bit pattern of `exp` — which is fine by
+    /// construction: the caller asserts the exponent is public.
+    ///
+    /// **Do not call with a secret exponent.** Use [`exp`](Self::exp)
+    /// instead, which is a fixed-iteration Montgomery ladder.
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/field.rs:575]: done

File: modmath/src/strict/montgomery.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// Constant-time finalize. Branchless conditional-subtract via
+    /// `subtle::ConditionallySelectable`.
+    pub mod ct {
+        #[doc(inline)]
+        pub use crate::montgomery::basic_mont::strict_wide_montgomery_mul_ct as mul;
+        #[doc(inline)]
+        pub use crate::montgomery::basic_mont::strict_wide_redc_ct as redc;
+    }
+}
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/strict/montgomery.rs:68]: done

File: modmath/src/constrained/mod.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+pub mod pre_reduced {
+    #[doc(inline)]
+    pub use crate::add::constrained_mod_add_pr as add;
+    #[doc(inline)]
+    pub use crate::sub::constrained_mod_sub_pr as sub;
+    #[doc(inline)]
+    pub use crate::mul::constrained_mod_mul_pr as mul;
+    #[doc(inline)]
+    pub use crate::exp::constrained_mod_exp_pr as exp;
+}
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/constrained/mod.rs:47]: done

File: modmath/src/basic/montgomery.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// Constant-time finalize. Branchless conditional-subtract via
+    /// `subtle::ConditionallySelectable`.
+    pub mod ct {
+        #[doc(inline)]
+        pub use crate::montgomery::basic_mont::wide_montgomery_mul_ct as mul;
+        #[doc(inline)]
+        pub use crate::montgomery::basic_mont::wide_redc_ct as redc;
+    }
+}
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/basic/montgomery.rs:103]: done

File: modmath/Cargo.toml [hunk 6]
```diff
@@ -16,10 +16,11 @@ keywords = ["numerics", "bignum"]
 num-traits = { version = "0.2", default-features = false }
 num-integer = { version = "0.1", default-features = false, optional = true }
 c0nst = { version = "0.2", optional = true }
-fixed-bigint = { version = "0.2.5", optional = true, default-features = false }
+fixed-bigint = { version = "0.3", optional = true, default-features = false }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/Cargo.toml:19]: https://crates.io/crates/fixed-bigint/0.3.0 exists

**Comment by kaidokert** [comment]
@sourcery-ai review

File: modmath/src/montgomery/basic_mont.rs [hunk 7]
```diff
@@ -457,7 +488,7 @@ where
 /// - Returns (result + carry1, extra_bit) where extra_bit indicates overflow
 fn accumulate_high_half_carry<T>(result: T, carry1: bool, carry2: bool) -> (T, bool)
 where
-    T: Copy + num_traits::One + num_traits::ops::overflowing::OverflowingAdd,
+    T: num_traits::One + num_traits::ops::overflowing::OverflowingAdd,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/basic_mont.rs:489]: @gemini-code-assist Addressed in fdbe317. Added `accumulate_high_half_carry_ct` as a constant-time sibling: branchless +1 via `subtle::ConditionallySelectable::conditional_select` driven by `Choice::from(carry1 as u8)`, bitwise carry combination on `subtle::Choice` (no `&&` / `||` short-circuit). NCT call sites keep the original helper; the two CT REDC call sites now use the new one. Exhaustive u8 × {false,true}² equivalence test confirms identical outputs.

File: modmath/src/montgomery/basic_mont.rs [hunk 8]
```diff
@@ -507,6 +545,105 @@ where
...
+        + subtle::ConditionallySelectable
+        + subtle::ConstantTimeLess,
+{
+    use subtle::Choice;
+
+    let m = t_lo.wrapping_mul(&n_prime);
+    let (m_lo, m_hi) = m.wide_mul(&modulus);
+    let (_discard_lo, carry1) = t_lo.overflowing_add(&m_lo);
+    let (result, carry2) = t_hi.overflowing_add(&m_hi);
+    let (result, extra_bit) = accumulate_high_half_carry(result, carry1, carry2);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/basic_mont.rs]: @gemini-code-assist Addressed in fdbe317 — `wide_redc_ct` now calls the constant-time `accumulate_high_half_carry_ct` helper instead.

File: modmath/src/montgomery/basic_mont.rs [hunk 9]
```diff
@@ -507,6 +545,105 @@ where
...
+        + subtle::ConditionallySelectable
+        + subtle::ConstantTimeLess,
+{
+    use subtle::Choice;
+
+    let m = t_lo.wrapping_mul(n_prime);
+    let (m_lo, m_hi) = m.wide_mul(modulus);
+    let (_discard_lo, carry1) = t_lo.overflowing_add(&m_lo);
+    let (result, carry2) = t_hi.overflowing_add(&m_hi);
+    let (result, extra_bit) = accumulate_high_half_carry(result, carry1, carry2);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/basic_mont.rs]: @gemini-code-assist Addressed in fdbe317 — `strict_wide_redc_ct` now calls the constant-time `accumulate_high_half_carry_ct` helper instead.

File: modmath/src/montgomery/cios.rs [hunk 10]
```diff
@@ -75,18 +83,81 @@ where
...
+
+    for i in 0..n {
+        let ai = a.get_word(i).into_option()?;
+
+        let carry = T::mul_acc_row(ai, b, &mut acc, zero);
+        let (sum, overflow) = acc_hi.overflowing_add(&carry);
+        acc_hi = sum;
+        if overflow {
+            acc_hi2 = acc_hi2 + one;
+        }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/cios.rs]: @gemini-code-assist Addressed in 2a1f877. Phase 1 overflow accumulation is now branchless: `let overflow_word = T::Word::conditional_select(&zero, &one, Choice::from(overflow as u8)); acc_hi2 = acc_hi2 + overflow_word;` — matches your suggested formulation. CT-vs-NCT equivalence tests (u8, u32x4) still pass.

File: modmath/src/montgomery/cios.rs [hunk 11]
```diff
@@ -75,18 +83,81 @@ where
...
+where
+    T: MulAccOps<GetWordOutput = subtle::CtOption<<T as MulAccOps>::Word>>
+        + ConstBorrowingSub
+        + subtle::ConditionallySelectable,
+    T::Word: num_traits::Zero
+        + num_traits::One
+        + num_traits::WrappingMul
+        + num_traits::ops::overflowing::OverflowingAdd
+        + core::ops::Add<Output = T::Word>
+        + subtle::ConstantTimeEq,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/cios.rs]: @gemini-code-assist Addressed in 2a1f877 — `subtle::ConditionallySelectable` added to the `T::Word` bound on `cios_montgomery_mul_ct`, enabling the branchless overflow accumulation at line 124.

File: modmath/src/montgomery/cios.rs [hunk 12]
```diff
@@ -99,6 +170,38 @@ where
...
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/cios.rs]: @gemini-code-assist Addressed in 2a1f877 — `subtle::ConditionallySelectable` mirrored on the `T::Word` bound of the `CiosMontMulCt` blanket impl to match the updated function signature.

[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/cios.rs]: @gemini-code-assist The suggested trait bound is already in place at HEAD (2a1f877) — `impl CiosMontMulCt for T` at modmath/src/montgomery/cios.rs:187-198 matches your suggestion block byte-for-byte. No further code change required; the thread can be marked resolved.
